File: lib/compiler/parser.js
/**
 * Parser
 * Converts tokens into an Abstract Syntax Tree (AST)
 * Uses recursive descent parsing
 */

import { TokenType } from './lexer.js';

// AST Node Types
export class ASTNode {
  constructor(type) {
    this.type = type;
  }
}

export class ProgramNode extends ASTNode {
  constructor(statements) {
    super('Program');
    this.statements = statements;
  }
}

export class VariableDeclarationNode extends ASTNode {
  constructor(identifier, initializer) {
    super('VariableDeclaration');
    this.identifier = identifier;
    this.initializer = initializer;
  }
}

export class AssignmentNode extends ASTNode {
  constructor(identifier, value) {
    super('Assignment');
    this.identifier = identifier;
    this.value = value;
  }
}

export class BinaryOpNode extends ASTNode {
  constructor(operator, left, right) {
    super('BinaryOp');
    this.operator = operator;
    this.left = left;
    this.right = right;
  }
}

export class UnaryOpNode extends ASTNode {
  constructor(operator, operand) {
    super('UnaryOp');
    this.operator = operator;
    this.operand = operand;
  }
}

export class NumberNode extends ASTNode {
  constructor(value) {
    super('Number');
    this.value = value;
  }
}

export class StringNode extends ASTNode {
  constructor(value) {
    super('String');
    this.value = value;
  }
}

export class IdentifierNode extends ASTNode {
  constructor(name) {
    super('Identifier');
    this.name = name;
  }
}

export class IfNode extends ASTNode {
  constructor(condition, thenBranch, elseBranch = null) {
    super('If');
    this.condition = condition;
    this.thenBranch = thenBranch;
    this.elseBranch = elseBranch;
  }
}

export class WhileNode extends ASTNode {
  constructor(condition, body) {
    super('While');
    this.condition = condition;
    this.body = body;
  }
}

export class BlockNode extends ASTNode {
  constructor(statements) {
    super('Block');
    this.statements = statements;
  }
}

export class PrintNode extends ASTNode {
  constructor(expression) {
    super('Print');
    this.expression = expression;
  }
}

export class Parser {
  constructor(tokens) {
    this.tokens = tokens.filter(t => t.type !== TokenType.NEWLINE); // Filter out newlines
    this.position = 0;
  }

  currentToken() {
    return this.tokens[this.position];
  }

  peek(offset = 1) {
    const pos = this.position + offset;
    return pos < this.tokens.length ? this.tokens[pos] : this.tokens[this.tokens.length - 1];
  }

  advance() {
    if (this.position < this.tokens.length - 1) {
      this.position++;
    }
    return this.currentToken();
  }

  expect(type) {
    const token = this.currentToken();
    if (token.type !== type) {
      throw new Error(
        `Expected token type ${type} but got ${token.type} at ${token.line}:${token.column}`
      );
    }
    this.advance();
    return token;
  }

  parse() {
    const statements = [];
    
    while (this.currentToken().type !== TokenType.EOF) {
      statements.push(this.parseStatement());
    }

    return new ProgramNode(statements);
  }

  parseStatement() {
    const token = this.currentToken();

    switch (token.type) {
      case TokenType.LET:
        return this.parseVariableDeclaration();
      case TokenType.IF:
        return this.parseIfStatement();
      case TokenType.WHILE:
        return this.parseWhileStatement();
      case TokenType.PRINT:
        return this.parsePrintStatement();
      case TokenType.LBRACE:
        return this.parseBlock();
      case TokenType.IDENTIFIER:
        return this.parseAssignmentOrExpression();
      default:
        throw new Error(`Unexpected token ${token.type} at ${token.line}:${token.column}`);
    }
  }

  parseVariableDeclaration() {
    this.expect(TokenType.LET);
    const identifier = this.expect(TokenType.IDENTIFIER).value;
    this.expect(TokenType.ASSIGN);
    const initializer = this.parseExpression();
    this.expect(TokenType.SEMICOLON);
    return new VariableDeclarationNode(identifier, initializer);
  }

  parseAssignmentOrExpression() {
    const identifier = this.currentToken().value;
    const token = this.currentToken();
    this.advance();

    if (this.currentToken().type === TokenType.ASSIGN) {
      this.advance();
      const value = this.parseExpression();
      this.expect(TokenType.SEMICOLON);
      return new AssignmentNode(identifier, value);
    } else {
      // Standalone identifier followed by semicolon is not a valid statement
      throw new Error(
        `Invalid statement: standalone identifier '${identifier}' at ${token.line}:${token.column}. ` +
        `Did you mean to assign a value or use it in an expression?`
      );
    }
  }

  parseIfStatement() {
    this.expect(TokenType.IF);
    this.expect(TokenType.LPAREN);
    const condition = this.parseExpression();
    this.expect(TokenType.RPAREN);
    const thenBranch = this.parseStatement();
    
    let elseBranch = null;
    if (this.currentToken().type === TokenType.ELSE) {
      this.advance();
      elseBranch = this.parseStatement();
    }

    return new IfNode(condition, thenBranch, elseBranch);
  }

  parseWhileStatement() {
    this.expect(TokenType.WHILE);
    this.expect(TokenType.LPAREN);
    const condition = this.parseExpression();
    this.expect(TokenType.RPAREN);
    const body = this.parseStatement();
    return new WhileNode(condition, body);
  }

  parsePrintStatement() {
    this.expect(TokenType.PRINT);
    this.expect(TokenType.LPAREN);
    const expression = this.parseExpression();
    this.expect(TokenType.RPAREN);
    this.expect(TokenType.SEMICOLON);
    return new PrintNode(expression);
  }

  parseBlock() {
    this.expect(TokenType.LBRACE);
    const statements = [];

    while (this.currentToken().type !== TokenType.RBRACE && 
           this.currentToken().type !== TokenType.EOF) {
      statements.push(this.parseStatement());
    }

    this.expect(TokenType.RBRACE);
    return new BlockNode(statements);
  }

  parseExpression() {
    return this.parseComparison();
  }

  parseComparison() {
    let left = this.parseTerm();

    while ([TokenType.EQUAL, TokenType.NOT_EQUAL, TokenType.LESS_THAN, 
            TokenType.GREATER_THAN, TokenType.LESS_EQUAL, TokenType.GREATER_EQUAL]
           .includes(this.currentToken().type)) {
      const operator = this.currentToken().value;
      this.advance();
      const right = this.parseTerm();
      left = new BinaryOpNode(operator, left, right);
    }

    return left;
  }

  parseTerm() {
    let left = this.parseFactor();

    while ([TokenType.PLUS, TokenType.MINUS].includes(this.currentToken().type)) {
      const operator = this.currentToken().value;
      this.advance();
      const right = this.parseFactor();
      left = new BinaryOpNode(operator, left, right);
    }

    return left;
  }

  parseFactor() {
    let left = this.parseUnary();

    while ([TokenType.MULTIPLY, TokenType.DIVIDE, TokenType.MODULO]
           .includes(this.currentToken().type)) {
      const operator = this.currentToken().value;
      this.advance();
      const right = this.parseUnary();
      left = new BinaryOpNode(operator, left, right);
    }

    return left;
  }

  parseUnary() {
    if ([TokenType.PLUS, TokenType.MINUS].includes(this.currentToken().type)) {
      const operator = this.currentToken().value;
      this.advance();
      const operand = this.parseUnary();
      return new UnaryOpNode(operator, operand);
    }

    return this.parsePrimary();
  }

  parsePrimary() {
    const token = this.currentToken();

    if (token.type === TokenType.NUMBER) {
      this.advance();
      return new NumberNode(token.value);
    }

    if (token.type === TokenType.STRING) {
      this.advance();
      return new StringNode(token.value);
    }

    if (token.type === TokenType.IDENTIFIER) {
      this.advance();
      return new IdentifierNode(token.value);
    }

    if (token.type === TokenType.LPAREN) {
      this.advance();
      const expr = this.parseExpression();
      this.expect(TokenType.RPAREN);
      return expr;
    }

    throw new Error(`Unexpected token ${token.type} at ${token.line}:${token.column}`);
  }
}

File: api/compiler/parser.py
import ply.yacc as yacc
from .lexer import tokens

precedence = (
    ('left', 'EQ', 'NE', 'LT', 'GT', 'LE', 'GE'),
    ('left', 'PLUS', 'MINUS'),
    ('left', 'TIMES', 'DIVIDE', 'MODULO'),
)

class ASTNode:
    def __init__(self, node_type, **attributes):
        self.type = node_type
        self.__dict__.update(attributes)
    
    def to_dict(self):
        result = {'type': self.type}
        
        for key, value in self.__dict__.items():
            if key == 'type':
                continue
                
            if isinstance(value, ASTNode):
                result[key] = value.to_dict()
            elif isinstance(value, list):
                result[key] = [item.to_dict() if isinstance(item, ASTNode) else item 
                              for item in value]
            else:
                result[key] = value
                
        return result


def p_program(p):
    """program : statement_list"""
    p[0] = ASTNode('Program', statements=p[1])


def p_statement_list(p):
    """statement_list : statement_list statement
                      | statement"""
    if len(p) == 3:
        p[0] = p[1] + [p[2]]
    else:
        p[0] = [p[1]]


def p_statement(p):
    """statement : declaration
                 | assignment
                 | if_statement
                 | while_statement
                 | print_statement"""
    p[0] = p[1]


def p_declaration(p):
    """declaration : LET ID ASSIGN expression SEMICOLON""" 
    p[0] = ASTNode('Declaration', name=p[2], value=p[4])


def p_assignment(p):
    """assignment : ID ASSIGN expression SEMICOLON"""
    p[0] = ASTNode('Assignment', name=p[1], value=p[3])


def p_if_statement(p):
    """if_statement : IF LPAREN expression RPAREN LBRACE statement_list RBRACE
                    | IF LPAREN expression RPAREN LBRACE statement_list RBRACE ELSE LBRACE statement_list RBRACE"""
    has_else = len(p) == 12
    p[0] = ASTNode('IfStatement', 
                   condition=p[3], 
                   then_branch=p[6],
                   else_branch=p[10] if has_else else None)


def p_while_statement(p):
    """while_statement : WHILE LPAREN expression RPAREN LBRACE statement_list RBRACE"""
    p[0] = ASTNode('WhileStatement', condition=p[3], body=p[6])


def p_print_statement(p):
    """print_statement : PRINT LPAREN expression RPAREN SEMICOLON"""
    p[0] = ASTNode('PrintStatement', value=p[3])


def p_expression_binop(p):
    """expression : expression PLUS expression
                  | expression MINUS expression
                  | expression TIMES expression
                  | expression DIVIDE expression
                  | expression MODULO expression
                  | expression EQ expression
                  | expression NE expression
                  | expression LT expression
                  | expression GT expression
                  | expression LE expression
                  | expression GE expression"""
    p[0] = ASTNode('BinaryOp', operator=p[2], left=p[1], right=p[3])


def p_expression_group(p):
    """expression : LPAREN expression RPAREN"""
    p[0] = p[2]


def p_expression_number(p):
    """expression : NUMBER"""
    p[0] = ASTNode('Number', value=p[1])


def p_expression_id(p):
    """expression : ID"""
    p[0] = ASTNode('Identifier', name=p[1])


def p_error(p):
    if p:
        error_msg = f"Syntax error at '{p.value}' (line {p.lineno})"
    else:
        error_msg = "Unexpected end of file"
    
    raise SyntaxError(error_msg)


def build_parser():
    # write_tables=False and debug=False prevent PLY from trying to write
    # parsetab.py / parser.out to the filesystem (required for serverless envs)
    return yacc.yacc(write_tables=False, debug=False)


def parse(code):
    from .lexer import build_lexer
    
    lexer = build_lexer()
    parser = build_parser()
    
    ast = parser.parse(code, lexer=lexer)
    
    return ast.to_dict() if ast else None
